src/libhidpp/hid/RawDevice.cpp

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2021 Clément Vuchener
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "RawDevice.h"
+
+#include <misc/Log.h>
+
+using namespace HID;
+
+void RawDevice::logReportDescriptor () const
+{
+	auto debug = Log::debug ("reportdesc");
+	if (!debug)
+		return;
+	for (const auto &collection: _report_desc.collections) {
+		debug << "Collection: " << std::hex << uint32_t (collection.usage) << std::dec << std::endl;
+		for (const auto &[id, fields]: collection.reports) {
+			const char *type;
+			switch (id.type) {
+			case ReportID::Type::Input: type = "Input"; break;
+			case ReportID::Type::Output: type = "Output"; break;
+			case ReportID::Type::Feature: type = "Feature"; break;
+			}
+			debug.printf ("- Report %s %d\n", type, id.id);
+			for (const auto &f: fields) {
+				debug.printf ("  - Flags: %x (%s, %s), Size: %d*%d\n",
+						f.flags.bits,
+						f.flags.Data () ? "Data" : "Constant",
+						f.flags.Array () ? "Array" : "Variable",
+						f.count, f.size);
+				struct {
```

Comment:
Are you aware of how `if constexpr` helps with variants? https://en.cppreference.com/w/cpp/utility/variant/visit has an example of that pattern which I find very helpful.

Reply:
I prefer to use separate functions if they don't share code. So either as I did or using `overloaded` (which is sadly not part of the standard library).